clang/tools/pattern-rewriter/src/pattern.cpp

```diff
@@ -0,0 +1,497 @@
+      }
+      Index++;
+      PatternIndex++;
+      continue;
+    }
+
+    if (std::holds_alternative<CodeElement>(Element)) {
+      const auto &Code = std::get<CodeElement>(Element);
+      MatchPattern Suffix(Pattern.begin() + PatternIndex + 1,
+                          Pattern.begin() + PatternIndex + 1 +
+                              Code.SuffixLength);
+
+      int Next = parseCodeElement(Suffix, Input, Index);
+      if (Next == -1) {
+        return {};
+      }
+      const int Indentation = detectIndentation(Input, Index);
+      std::string ElementContents =
+          dedent(Input.substr(Index, Next - Index), Indentation);
+      if (Result.Bindings.count(Code.Name)) {
+        if (Result.Bindings[Code.Name] != ElementContents) {
+          return {};
+        }
+      } else {
+        Result.Bindings[Code.Name] = std::move(ElementContents);
+      }
+      Index = Next;
+      PatternIndex++;
+      continue;
+    }
+
+    PatternIndex++;
+  }
+
+  Result.Start = Start;
+  Result.End = Index;
+  return Result;
+}
+
+static void instantiateTemplate(
+    const std::string &Template,
+    const std::unordered_map<std::string, std::string> &Bindings,
+    const int Indentation, std::ostream &OutputStream) {
+  const int Size = Template.size();
+  int Index = 0;
+
+  while (Index < Size && isWhitespace(Template[Index])) {
+    Index++;
+  }
+
+  int End = Size - 1;
+  while (End > (Index + 1) && isWhitespace(Template[End])) {
+    End--;
+  }
+
+  while (Index <= End) {
+    const auto Character = Template[Index];
+
+    if (Index < (Size - 1) && Character == '$' && Template[Index + 1] == '{') {
+      const int BindingStart = Index;
+      Index += 2;
+
+      std::stringstream NameStream;
+      while (Index < Size && Template[Index] != '}') {
+        NameStream << Template[Index];
+        Index++;
+      }
+
+      if (Index == Size) {
+        throw std::runtime_error("Invalid rewrite pattern expression");
+      }
+      Index++;
+
+      std::string Name = NameStream.str();
+      if (Bindings.count(Name) > 0) {
+        const int BindingIndentation =
+            detectIndentation(Template, BindingStart) + Indentation;
+        const std::string Contents =
+            indent(Bindings.at(Name), BindingIndentation);
+        OutputStream << Contents;
+      }
+      continue;
+    }
+
+    OutputStream << Character;
+    if (Character == '\n') {
+      for (int i = 0; i < Indentation; i++)
+        OutputStream << " ";
+    }
+
+    Index++;
+  }
+}
+
+std::string applyRule(const Rule &Rule, const std::string &Input) {
+  std::stringstream OutputStream;
+  const auto Pattern = parseMatchPattern(Rule.Match);
+  const int Size = Input.size();
+
+  int Index = 0;
+  while (Index < Size) {
+    auto Result = findMatch(Pattern, Input, Index);
+
+    if (Result.has_value()) {
+      auto &Match = Result.value();
+      for (const auto &[Name, Value] : Match.Bindings) {
+        if (Rule.Subrules.count(Name)) {
```

Comment:
the combination of count() + at() can be replaced by find()
can save one of the search time